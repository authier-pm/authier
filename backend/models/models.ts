import { prisma } from '../prisma'
import {
  Arg,
  Ctx,
  Field,
  InputType,
  Int,
  ObjectType,
  UseMiddleware
} from 'type-graphql'

import { EncryptedSecrets, User } from '../generated/typegraphql-prisma'

@ObjectType()
export class UserBase {
  @Field(() => String)
  id: string

  @Field(() => String, { nullable: true })
  email?: string

  @Field(() => String)
  phone_number?: string

  @Field(() => String)
  account_name?: string

  @Field(() => String)
  password: string

  @Field(() => Number)
  tokenVersion: number

  @Field(() => Int)
  primaryDeviceId: number
}

@ObjectType()
export class UserAfterAuth extends User {
  @Field(() => [EncryptedSecrets], { nullable: true })
  EncryptedSecrets: Array<EncryptedSecrets> | undefined
}

@ObjectType()
export class LoginResponse {
  @Field(() => String)
  accessToken: string

  @Field(() => UserAfterAuth)
  user: UserAfterAuth
}

@InputType()
export class OTPEvent {
  @Field(() => String)
  kind: string

  @Field(() => String)
  url: string
}
