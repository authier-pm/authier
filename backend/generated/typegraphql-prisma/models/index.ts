export { Device } from './Device'
export { EncryptedSecrets } from './EncryptedSecrets'
export { EncryptedSecretsChangeAction } from './EncryptedSecretsChangeAction'
export { OTPCodeEvent } from './OTPCodeEvent'
export { SettingsConfig } from './SettingsConfig'
export { Tag } from './Tag'
export { Token } from './Token'
export { User } from './User'
export { UserPaidProducts } from './UserPaidProducts'
export { VaultUnlockEvents } from './VaultUnlockEvents'
export { WebInput } from './WebInput'
