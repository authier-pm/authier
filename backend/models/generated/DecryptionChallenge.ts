import { Field, ObjectType, Int } from 'type-graphql'
import { UserGQL } from './User'
import { DeviceGQL } from './Device'

@ObjectType()
export class DecryptionChallengeGQLScalars {
  @Field(() => Int)
  id: number

  @Field({ nullable: true })
  masterPasswordVerifiedAt?: Date

  @Field({ nullable: true })
  approvedAt?: Date

  @Field({ nullable: true })
  rejectedAt?: Date

  @Field({ nullable: true })
  blockIp?: boolean

  @Field()
  deviceName: string

  @Field()
  userId: string

  @Field()
  createdAt: Date

  @Field()
  deviceId: string

  @Field()
  approvedByRecovery: boolean

  @Field({ nullable: true })
  approvedFromDeviceId?: string
}

@ObjectType()
export class DecryptionChallengeGQL extends DecryptionChallengeGQLScalars {
  @Field(() => UserGQL)
  user: UserGQL

  @Field(() => DeviceGQL, { nullable: true })
  approvedFromDevice?: DeviceGQL

  @Field(() => [UserGQL])
  User: UserGQL[]

  // skip overwrite ğŸ‘‡
}
