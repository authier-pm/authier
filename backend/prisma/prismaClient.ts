import 'dotenv/config'

import debug from 'debug'
import { enablePrismaDebug } from './prismaDebug'
import { getDbCount } from '../scripts/getDbCount'

import { PrismaClient, Prisma } from '@prisma/client'
import kyselyExtension from 'prisma-extension-kysely'
import type { DB } from './generated/types'

import {
  Kysely,
  PostgresAdapter,
  PostgresIntrospector,
  PostgresQueryCompiler
} from 'kysely'

const log = debug('prisma:sql')
const logQueries = debug('au:prisma')

const NODE_ENV = process.env.NODE_ENV || 'test'

let dbUrl = process.env.DATABASE_URL
console.log('~ dbUrl', dbUrl)

const workerId = process.env.VITEST_WORKER_ID
if (workerId) {
  dbUrl = dbUrl?.includes('?') ? dbUrl?.split('?')[0] : dbUrl
  const vitestWorkerId = Number(process.env.VITEST_WORKER_ID) % getDbCount()
  dbUrl = `${dbUrl}_test_${
    vitestWorkerId + 1
  }?connection_limit=500&pool_timeout=0&connect_timeout=0` // this allows us to run tests in parallel against multiple dbs without conflicts
} else {
  log('DATABASE_URL', dbUrl)
}

export const prismaClient = new PrismaClient({
  transactionOptions: {
    timeout: 55_000,
    maxWait: 55_000
  },
  log:
    NODE_ENV === 'production'
      ? ['info', 'warn']
      : [
          {
            emit: 'event',
            level: 'query'
          },
          'info',
          'warn'
        ],
  errorFormat: workerId ? 'pretty' : undefined,
  datasources: {
    db: {
      url: dbUrl
    }
  }
}).$extends(
  kyselyExtension({
    kysely: (driver) =>
      new Kysely<DB>({
        dialect: {
          // This is where the magic happens!
          createDriver: () => driver,
          // Don't forget to customize these to match your database!
          createAdapter: () => new PostgresAdapter(),
          createIntrospector: (db) => new PostgresIntrospector(db),
          createQueryCompiler: () => new PostgresQueryCompiler()
        }
      })
  })
)

const debugLogs = process.env.LOG_PRISMA_SQL

if (debugLogs) {
  enablePrismaDebug(prismaClient)
}

// @ts-expect-error
export const dmmf = prismaClient._runtimeDataModel as any
