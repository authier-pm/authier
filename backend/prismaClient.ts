import { Prisma, PrismaClient } from '@prisma/client'
import './dotenv'

const nodeEnv = process.env.NODE_ENV || 'test'

const logConfig =
  nodeEnv === 'production'
    ? (['info', 'warn'] as Array<Prisma.LogLevel>)
    : [
        {
          emit: 'event',
          level: 'query'
        } as Prisma.LogDefinition,
        'info' as Prisma.LogLevel,
        'warn' as Prisma.LogLevel
      ]

export const prismaClient = new PrismaClient({
  log: logConfig
})
