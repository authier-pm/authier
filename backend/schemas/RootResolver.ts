import { throwIfNotAuthenticated } from '../api/authMiddleware'
import {
  Query,
  Resolver,
  Mutation,
  Arg,
  Ctx,
  UseMiddleware,
  Info,
  Int
} from 'type-graphql'
import { dmmf, prismaClient } from '../prisma/prismaClient'
import { FastifyReply, FastifyRequest } from 'fastify'
import { LoginResponse } from '../models/models'

import { verify } from 'jsonwebtoken'
import { UserQuery } from '../models/UserQuery'
import { UserMutation } from '../models/UserMutation'
import { constructURL } from '../../shared/urlUtils'

import { GraphqlError } from '../api/GraphqlError'
import { WebInputElement } from '../models/WebInputElement'
import { GraphQLEmailAddress, GraphQLUUID } from 'graphql-scalars'
import debug from 'debug'
import { RegisterNewAccountInput } from '../models/AuthInputs'

import { Device, User, WebInput } from '.prisma/client'

import { WebInputGQL } from '../models/generated/WebInputGQL'

import { GraphQLResolveInfo } from 'graphql'
import { getPrismaRelationsFromGQLInfo } from '../utils/getPrismaRelationsFromInfo'

import { DeviceInput, DeviceMutation, DeviceQuery } from '../models/Device'
import {
  DecryptionChallengeApproved,
  DecryptionChallengeForApproval,
  DecryptionChallengeUnion
} from '../models/DecryptionChallenge'
import { plainToClass } from 'class-transformer'
import type { PrismaClientKnownRequestError } from '@prisma/client/runtime'
import admin from 'firebase-admin'
import { isTorExit } from './isTorExit'
const log = debug('au:RootResolver')

export interface IContext {
  request: FastifyRequest
  reply: FastifyReply
  getIpAddress: () => string
  prisma: typeof prismaClient
}

export interface IJWTPayload {
  userId: string
  deviceId: string
}

export interface IContextAuthenticated extends IContext {
  jwtPayload: IJWTPayload
  device: Device
  masterDeviceId: string | null | undefined
}

export interface IContextMaybeAuthenticated extends IContext {
  jwtPayload?: IJWTPayload
}

export interface Payload {
  userId: string
  iat: number
  exp: number
}

@Resolver()
export class RootResolver {
  @Query(() => String)
  osTime() {
    return new Date().toISOString()
  }

  @Query(() => Boolean, {
    description: 'you need to be authenticated to call this resolver'
  })
  authenticated(@Ctx() ctx: IContext) {
    const inCookies = ctx.request.cookies['access-token']
    const inHeader = ctx.request.headers['authorization']

    try {
      if (inHeader) {
        const token = inHeader?.split(' ')[1]
        verify(token, process.env.ACCESS_TOKEN_SECRET!)
        return true
      } else if (inCookies) {
        verify(inCookies, process.env.ACCESS_TOKEN_SECRET!)
        return true
      }

      return false
    } catch (err) {
      return false
    }
  }

  @UseMiddleware(throwIfNotAuthenticated)
  @Query(() => UserQuery)
  @Mutation(() => UserMutation, {
    description: 'you need to be authenticated to call this resolver',
    name: 'me',
    nullable: false
  })
  async me(
    @Ctx() ctx: IContextAuthenticated,
    @Info() info: GraphQLResolveInfo
  ) {
    const { jwtPayload } = ctx

    const include = getPrismaRelationsFromGQLInfo({
      info,
      rootModel: dmmf.models.User
    })

    const tmp = await ctx.prisma.user.findUnique({
      where: { id: jwtPayload.userId },
      include
    })

    return tmp
  }

  @UseMiddleware(throwIfNotAuthenticated)
  @Query(() => DeviceQuery)
  @Mutation(() => DeviceMutation)
  async currentDevice(
    @Ctx() ctx: IContextAuthenticated,
    @Info() info: GraphQLResolveInfo
  ) {
    const { jwtPayload } = ctx

    return ctx.prisma.device.findUnique({
      where: { id: jwtPayload.deviceId },
      include: getPrismaRelationsFromGQLInfo({
        info,
        rootModel: dmmf.models.Device
      })
    })
  }

  @Mutation(() => LoginResponse)
  async registerNewUser(
    @Arg('input', () => RegisterNewAccountInput) input: RegisterNewAccountInput,
    @Arg('userId', () => GraphQLUUID) userId: string,
    @Ctx() ctx: IContext
  ) {
    const ipAddress = ctx.getIpAddress()
    const {
      email,
      firebaseToken,
      deviceName,
      deviceId,
      addDeviceSecret,
      addDeviceSecretEncrypted,
      encryptionSalt
    } = input
    let user: User & { Devices: Device[] }

    try {
      user = await ctx.prisma.user.create({
        data: {
          id: userId,
          email: email,
          addDeviceSecret,
          addDeviceSecretEncrypted,
          encryptionSalt,
          deviceRecoveryCooldownMinutes: 16 * 60, // 16 hours should be plenty enough
          loginCredentialsLimit: 40, // default limit
          TOTPlimit: 3,
          Devices: {
            create: {
              syncTOTP: true,
              platform: input.devicePlatform,
              id: deviceId,
              firstIpAddress: ipAddress,
              lastIpAddress: ipAddress,
              firebaseToken: firebaseToken,
              name: deviceName
            }
          }
        },
        include: {
          EncryptedSecrets: true,
          Devices: true
        }
      })
    } catch (err: PrismaClientKnownRequestError | any) {
      if (err.code === 'P2002' && err.meta.target[0] === 'email') {
        log('email', email)

        throw new GraphqlError(`User with such email already exists.`)
      }
      if (err.code === 'P2002' && err.meta.target[0] === 'id') {
        log('deviceId', deviceId)
        if (process.env.NODE_ENV === 'development') {
          console.warn(
            `deleting device ${deviceId} because we are in dev mode and we don't care about the other account`
          )
          await prismaClient.device.delete({
            where: { id: deviceId }
          })
          return this.registerNewUser(input, userId, ctx)
        } else {
          throw new GraphqlError(
            `Device ${deviceId} already exists. You cannot use a device with multiple accounts.`
          )
        }
      }
      throw err
    }

    const device = user.Devices[0]
    user = await ctx.prisma.user.update({
      where: {
        id: user.id
      },
      data: {
        masterDeviceId: device.id
      },
      include: {
        Devices: true
      }
    })

    return new UserMutation(user).setCookiesAndConstructLoginResponse(
      device.id,
      ctx
    )
  }

  // TODO rate limit this per IP
  @Mutation(() => DecryptionChallengeUnion, {
    description: 'returns a decryption challenge',
    nullable: true
  })
  async deviceDecryptionChallenge(
    @Arg('email', () => GraphQLEmailAddress) email: string,
    @Arg('deviceInput', () => DeviceInput)
    deviceInput: DeviceInput,
    @Ctx() ctx: IContext
  ) {
    const ipAddress = ctx.getIpAddress()

    if (await isTorExit(ipAddress)) {
      throw new GraphqlError(
        'Tor exit nodes are prohibited from login/adding new devices.'
      )
    }

    const user = await ctx.prisma.user.findUnique({
      where: { email },
      select: {
        id: true,
        addDeviceSecretEncrypted: true,
        encryptionSalt: true,
        masterDevice: true
      }
    })

    if (!user) {
      throw new GraphqlError('Login failed, create a new account.')
    }
    const isBlocked = await ctx.prisma.decryptionChallenge.findFirst({
      where: {
        userId: user.id,
        blockIp: true,
        ipAddress
      }
    })

    if (isBlocked) {
      throw new GraphqlError('Login failed, try again later.')
    }

    const inLastHour = await ctx.prisma.decryptionChallenge.count({
      where: {
        userId: user.id,
        createdAt: {
          gte: new Date(Date.now() - 3600000)
        },
        masterPasswordVerifiedAt: null
      }
    })

    if (inLastHour > 5) {
      throw new GraphqlError(
        'Too many decryption challenges, wait for cooldown'
      )
    }

    const device = await ctx.prisma.device.findUnique({
      where: { id: deviceInput.id }
    })

    console.log('device', device)

    let challenge = await ctx.prisma.decryptionChallenge.findFirst({
      where: {
        deviceId: deviceInput.id,
        userId: user.id
      }
    })

    //TODO: Check this condition, not sure what is this doing
    if (device) {
      if (!challenge) {
        const deviceCount = await ctx.prisma.device.count({
          where: {
            userId: user.id
          }
        })
        if (deviceCount === 1) {
          // user has only one device
          challenge = await ctx.prisma.decryptionChallenge.create({
            data: {
              deviceId: deviceInput.id,
              deviceName: deviceInput.name,
              userId: user.id,
              ipAddress,
              approvedAt: device.createdAt // we mark this as approved immediately, because user has only one device
            }
          })
        }
      }
    }

    if (challenge?.rejectedAt) {
      // someone tried to login with this device and it was rejected in the past, we don't want to create a new challenge
      throw new GraphqlError('login failed')
    }

    if (!challenge) {
      // TODO: Will we have notifications for browser?
      if (
        user.masterDevice?.firebaseToken &&
        user.masterDevice.firebaseToken.length > 10
      ) {
        console.log('sending notification to')
        await admin.messaging().sendToDevice(user.masterDevice.firebaseToken, {
          notification: {
            title: 'New device login!',
            body: 'New device is trying to log in.'
          },
          data: {
            type: 'Devices'
          }
        })
      }
      challenge = await ctx.prisma.decryptionChallenge.create({
        data: {
          deviceId: deviceInput.id,
          deviceName: deviceInput.name,
          userId: user.id,
          ipAddress: ctx.getIpAddress()
        }
      })
    }

    if (!challenge.approvedAt) {
      // TODO: enable when we have device management in the vault
      return plainToClass(DecryptionChallengeForApproval, {
        id: challenge.id,
        approvedAt: challenge.approvedAt
      })
    }

    // user has approved this device in the past, we can return the challenge including salt and encrypted secret
    return plainToClass(DecryptionChallengeApproved, {
      ...challenge,

      addDeviceSecretEncrypted: user.addDeviceSecretEncrypted,
      encryptionSalt: user.encryptionSalt
    })
  }

  @UseMiddleware(throwIfNotAuthenticated)
  @Mutation(() => Int, {
    nullable: true,
    deprecationReason: 'prefer device methods',
    description:
      'removes current device. Returns null if user is not authenticated, alias for device logout/remove methods'
  })
  async logout(
    @Ctx() ctx: IContextAuthenticated,
    @Arg('removeDevice', { nullable: true }) removeDevice: boolean
  ) {
    ctx.reply.clearCookie('refresh-token')
    ctx.reply.clearCookie('access-token')

    if (!ctx.jwtPayload) {
      return null
    }
    const user = await ctx.prisma.user.update({
      data: {
        Devices: {
          update: {
            where: {
              id: ctx.jwtPayload.deviceId
            },
            data: {
              logoutAt: new Date()
            }
          }
        }
      },
      where: {
        id: ctx.jwtPayload.userId
      }
    })

    if (removeDevice) {
      await ctx.prisma.$transaction([
        ctx.prisma.device.delete({
          where: {
            id: ctx.jwtPayload.deviceId
          }
        }),
        ctx.prisma.decryptionChallenge.deleteMany({
          where: {
            deviceId: ctx.jwtPayload.deviceId
          }
        })
      ])
    }
    return user.tokenVersion
  }

  @Query(() => [WebInputGQL])
  async webInputs(
    @Arg('host') host: string,
    @Ctx() ctx: IContextAuthenticated
  ) {
    return ctx.prisma.webInput.findMany({ where: { host } })
  }

  @UseMiddleware(throwIfNotAuthenticated)
  @Mutation(() => [WebInputGQL])
  async addWebInputs(
    @Arg('webInputs', () => [WebInputElement]) webInputs: WebInputElement[],
    @Ctx() ctx: IContextAuthenticated
  ) {
    const returnedInputs: WebInput[] = []
    for (const webInput of webInputs) {
      const forUpsert = {
        url: webInput.url,
        host: constructURL(webInput.url).host,
        domPath: webInput.domPath,
        kind: webInput.kind,
        addedByUserId: ctx.jwtPayload.userId
      }
      const input = await ctx.prisma.webInput.upsert({
        create: forUpsert,
        update: forUpsert,
        where: {
          webInputIdentifier: {
            url: webInput.url,
            domPath: webInput.domPath
          }
        }
      })
      // @ts-ignore TODO figure out why this errors only on local, but not on CI
      returnedInputs.push(input)
    }
    return returnedInputs
  }
}
