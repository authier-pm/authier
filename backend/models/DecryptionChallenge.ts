import {
  Arg,
  Ctx,
  Field,
  Info,
  InputType,
  Int,
  Mutation,
  ObjectType
} from 'type-graphql'
import { IContext, IContextAuthenticated } from '../schemas/RootResolver'
import { DecryptionChallengeGQL } from './generated/DecryptionChallenge'
import { GraphQLResolveInfo } from 'graphql'
import { createUnionType } from 'type-graphql'
import { GraphQLNonEmptyString, GraphQLUUID } from 'graphql-scalars'
import { GraphqlError } from '../api/GraphqlError'
import { dmmf } from '../prisma/prismaClient'
import { getPrismaRelationsFromInfo } from '../utils/getPrismaRelationsFromInfo'
import { AddNewDeviceInput } from './AuthInputs'
import { LoginResponse } from './models'
import { UserMutation } from './UserMutation'

@InputType()
export class DeviceInput {
  @Field(() => GraphQLUUID)
  id: string

  @Field()
  name: string
}

@ObjectType()
export class DecryptionChallengeForApproval {
  @Field(() => Int)
  id: number

  @Field({ nullable: true })
  approvedAt?: Date

  @Field({ nullable: true })
  rejectedAt?: Date

  @Field()
  createdAt: Date

  @Field()
  deviceName: string

  @Field(() => GraphQLUUID)
  deviceId: string
}

@ObjectType()
export class DecryptionChallengeApproved extends DecryptionChallengeGQL {
  @Field()
  addDeviceSecretEncrypted: string

  @Field()
  encryptionSalt: string

  @Field(() => LoginResponse)
  async addNewDeviceForUser(
    @Arg('input', () => AddNewDeviceInput) input: AddNewDeviceInput,
    @Arg('currentAddDeviceSecret', () => GraphQLNonEmptyString)
    currentAddDeviceSecret: string,
    @Ctx() ctx: IContext,
    @Info() info: GraphQLResolveInfo
  ) {
    // const include = getPrismaRelationsFromInfo({
    //   info,
    //   rootModel: dmmf.modelMap.User
    // })

    const { id, deviceId, userId } = this
    const where = { id: userId }

    // TODO use findUnique when prisma bug gets fixed
    const user = await ctx.prisma.user.findFirst({
      where: where,
      include: {
        EncryptedSecrets: true
      }
    })

    if (!user) {
      throw new GraphqlError('User not found')
    }

    if (user?.addDeviceSecret !== currentAddDeviceSecret) {
      // TODO rate limit these attempts and notify current devices
      throw new GraphqlError('Wrong master password used')
    }

    await ctx.prisma.user.update({
      data: {
        addDeviceSecret: input.addDeviceSecret,
        addDeviceSecretEncrypted: input.addDeviceSecretEncrypted
      },
      where: {
        id: user.id
      }
    })

    await ctx.prisma.decryptionChallenge.updateMany({
      where: {
        id,
        deviceId,
        userId: user.id
      },
      data: { masterPasswordVerifiedAt: new Date() }
    })

    const { firebaseToken } = input
    const ipAddress = ctx.getIpAddress()

    let device = await ctx.prisma.device.findUnique({
      // TODO change this to upsert
      where: { id: deviceId }
    })

    if (device) {
      if (device.userId !== user.id) {
        throw new GraphqlError('Device is already registered for another user')
      }

      device = await ctx.prisma.device.update({
        data: { logoutAt: null },
        where: { id: device.id }
      })
    } else {
      device = await ctx.prisma.device.create({
        data: {
          id: deviceId,
          firstIpAddress: ipAddress,
          lastIpAddress: ipAddress,
          firebaseToken: firebaseToken,
          name: this.deviceName,
          userId: user.id,
          platform: 'chrome' // TODO add this to input
        }
      })
    }

    return new UserMutation(user).setCookiesAndConstructLoginResponse(
      device.id,
      ctx
    )
  }
}

@ObjectType()
export class DecryptionChallengeMutation extends DecryptionChallengeGQL {
  @Field(() => DecryptionChallengeGQL)
  async approve(@Ctx() ctx: IContextAuthenticated) {
    return ctx.prisma.decryptionChallenge.update({
      where: { id: this.id },
      data: {
        approvedAt: new Date(),
        rejectedAt: null,
        blockIp: this.blockIp ? false : null // if it was previously rejected, we mark it as false
      }
    })
  }

  @Field(() => DecryptionChallengeGQL)
  async reject(@Ctx() ctx: IContextAuthenticated) {
    return ctx.prisma.decryptionChallenge.update({
      where: { id: this.id },
      data: { rejectedAt: new Date(), blockIp: true, approvedAt: null }
    })
  }

  @Field(() => DecryptionChallengeGQL)
  async recoverAccount(@Ctx() ctx: IContextAuthenticated) {
    // TODO send notification to all contacts we have, just email for now
    return ctx.prisma.user.update({
      where: { id: this.userId },
      data: { recoveryDecryptionChallengeId: this.id } // rest is handled by our CRON job
    })
  }
}

export const DecryptionChallengeUnion = createUnionType({
  name: 'DecryptionChallenge', // the name of the GraphQL union
  types: () =>
    [DecryptionChallengeApproved, DecryptionChallengeForApproval] as const
})
