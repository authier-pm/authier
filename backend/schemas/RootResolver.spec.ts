import { prismaClient } from '../prisma/prismaClient'
import { IContextAuthenticated, RootResolver } from './RootResolver'
import faker, { fake } from 'faker'
import {
  AddNewDeviceInput,
  RegisterNewAccountInput
} from '../models/AuthInputs'
import { GraphQLResolveInfo } from 'graphql'

import { sign } from 'jsonwebtoken'
import { makeFakeCtx } from '../tests/makeFakeCtx'
import { DecryptionChallengeApproved } from '../models/DecryptionChallenge'

export const makeAddNewDeviceInput = () => ({
  email: faker.internet.email(),
  deviceName: faker.internet.userName(),
  deviceId: faker.datatype.uuid(),
  firebaseToken: faker.datatype.uuid(),
  addDeviceSecret: faker.datatype.string(5),
  addDeviceSecretEncrypted: faker.datatype.string(5),
  decryptionChallengeId: faker.datatype.number(),
  deviceRecoveryCooldownMinutes: faker.datatype.number()
})

export const makeRegisterAccountInput = () => ({
  ...makeAddNewDeviceInput(),
  encryptionSalt: faker.datatype.string(5)
})

const userSecurityProps = {
  deviceRecoveryCooldownMinutes: 960,
  loginCredentialsLimit: 50,
  TOTPlimit: 4
}

describe('RootResolver', () => {
  const resolver = new RootResolver()

  describe('me', () => {
    it('should return current user', async () => {
      const user = await prismaClient.user.create({
        data: {
          id: faker.datatype.uuid(),
          email: faker.internet.email(),
          addDeviceSecret: faker.datatype.string(5),
          addDeviceSecretEncrypted: faker.datatype.string(5),
          encryptionSalt: faker.datatype.string(5),
          ...userSecurityProps
        }
      })

      expect(
        await resolver.me(
          {
            request: { headers: {} },
            prisma: prismaClient,
            jwtPayload: { userId: user.id }
          } as IContextAuthenticated,
          {} as any
        )
      ).toMatchObject(user)
    })
  })

  describe('logout', () => {
    it('should clear both cookies', async () => {})
    it('should increment tokenVersion for the current user', async () => {})
  })

  describe('registerNewUser', () => {
    beforeEach(() => {
      jest.setTimeout(10000)
    })
    const userId = faker.datatype.uuid()

    it('should add new user', async () => {
      let input: RegisterNewAccountInput = makeRegisterAccountInput()

      const accessToken = sign(
        { userId: userId, deviceId: input.deviceId },
        process.env.ACCESS_TOKEN_SECRET!,
        {
          expiresIn: '60m'
        }
      )

      let data = await resolver.registerNewUser(
        input,
        userId,
        makeFakeCtx(userId)
      )

      expect({
        accessToken: data.accessToken,
        email: data.user.email
      }).toMatchObject({ accessToken: accessToken, email: input.email })

      await prismaClient.settingsConfig.deleteMany()
      await prismaClient.device.deleteMany()
      await prismaClient.user.deleteMany()
    })

    it('should throw User with such email already exists', async () => {
      let input: RegisterNewAccountInput = makeRegisterAccountInput()
      await prismaClient.user.create({
        data: {
          id: faker.datatype.uuid(),
          email: input.email,
          addDeviceSecret: input.addDeviceSecret,
          addDeviceSecretEncrypted: input.addDeviceSecretEncrypted,
          encryptionSalt: input.encryptionSalt,
          ...userSecurityProps
        }
      })

      await expect(
        async () =>
          await resolver.registerNewUser(input, userId, makeFakeCtx(userId))
      ).rejects.toThrow('User with such email already exists.')
    })

    it("should show 'Device already exists. You cannot register this device for multiple accounts.'", async () => {
      let userId = faker.datatype.uuid()

      let input: RegisterNewAccountInput = makeRegisterAccountInput()
      await prismaClient.user.create({
        data: {
          id: userId,
          email: faker.internet.email(),
          addDeviceSecret: input.addDeviceSecret,
          addDeviceSecretEncrypted: input.addDeviceSecretEncrypted,
          encryptionSalt: input.encryptionSalt,
          ...userSecurityProps,

          Devices: {
            create: {
              id: input.deviceId,
              firstIpAddress: faker.internet.ip(),
              lastIpAddress: faker.internet.ip(),
              firebaseToken: faker.datatype.uuid(),
              name: input.deviceName
            }
          }
        }
      })

      await expect(
        async () =>
          await resolver.registerNewUser(
            input,
            faker.datatype.uuid(),
            makeFakeCtx(userId)
          )
      ).rejects.toThrow(
        'Device already exists. You cannot register this device for multiple accounts.'
      )
    })
  })

  describe('deviceDecryptionChallenge', () => {
    it('should return a decryption challenge', async () => {
      let userId = faker.datatype.uuid()
      let fakeData: RegisterNewAccountInput = makeRegisterAccountInput()
      await prismaClient.user.create({
        data: {
          id: userId,
          email: fakeData.email,
          addDeviceSecret: fakeData.addDeviceSecret,
          addDeviceSecretEncrypted: fakeData.addDeviceSecretEncrypted,

          encryptionSalt: fakeData.encryptionSalt,
          ...userSecurityProps
        }
      })

      let data = (await resolver.deviceDecryptionChallenge(
        fakeData.email,
        faker.datatype.uuid(),
        makeFakeCtx(userId)
      )) as DecryptionChallengeApproved

      expect(data?.addDeviceSecretEncrypted).toBe(
        fakeData.addDeviceSecretEncrypted
      )
    })

    it("should show 'Too many decryption challenges, wait for cooldown'", async () => {
      let userId = faker.datatype.uuid()

      let fakeData: RegisterNewAccountInput = makeRegisterAccountInput()
      await prismaClient.user.create({
        data: {
          id: userId,
          email: fakeData.email,
          addDeviceSecret: fakeData.addDeviceSecret,
          addDeviceSecretEncrypted: fakeData.addDeviceSecretEncrypted,
          encryptionSalt: fakeData.encryptionSalt,

          ...userSecurityProps
        }
      })

      const data = Array.from({ length: 10 }).map(() => ({
        userId: userId,
        ipAddress: faker.internet.ip(),
        masterPasswordVerifiedAt: null
      }))

      await prismaClient.decryptionChallenge.createMany({
        data
      })

      await expect(async () => {
        await resolver.deviceDecryptionChallenge(
          fakeData.email,
          faker.datatype.uuid(),
          makeFakeCtx(userId)
        )
      }).rejects.toThrow('Too many decryption challenges, wait for cooldown')
    })

    it.todo(
      'should block creation of a challenge from an IP which was blocked previously'
    )
  })
})
