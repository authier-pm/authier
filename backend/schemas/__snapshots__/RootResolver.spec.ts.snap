// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RootResolver > addWebInputs > should add to the DB and omit any url query when storing the url 1`] = `
{
  "addedByUserId": "0fff9a02-7a89-487e-86a9-1d4c796e517f",
  "domOrdinal": 0,
  "domPath": "body",
  "host": "google.com",
  "kind": "PASSWORD",
  "layoutType": null,
  "url": "https://google.com",
}
`;
